clang/test/CodeGenSYCL/device_global.cpp

```diff
@@ -124,8 +124,250 @@ void foo() {
+// CHECK: @counter15 = addrspace(1) global %"class.sycl::_V1::ext::oneapi::device_global.18" zeroinitializer, align 8 #[[DEV_GLOB_FPGA_ATTRS15:[0-9]+]]
+// CHECK: [[ANN_private_copies_imple_dual_port2:@.str[.0-9]*]] = {{.*}}{memory:DEFAULT}{sizeinfo:4}{simple_dual_port:1}{memory:DEFAULT}{sizeinfo:4}{private_copies:16}{{.*}}
+// CHECK: [[ANN_numbanks3_merge3:@.str[.0-9]*]] = {{.*}}{memory:DEFAULT}{sizeinfo:4}{numbanks:2}{memory:DEFAULT}{sizeinfo:4,155}{merge:foo:depth}{{.*}}
+// CHECK: [[ANN_force_pow2_depth3_doublepump3_numbanks3:@.str[.0-9]*]]  = {{.*}}{memory:DEFAULT}{sizeinfo:4}{force_pow2_depth:0}{memory:DEFAULT}{sizeinfo:4}{pump:2}{memory:DEFAULT}{sizeinfo:4,155}{numbanks:4}{bank_bits:2,3}{{.*}}
 // CHECK: @_ZN12_GLOBAL__N_19same_nameE = internal addrspace(1) global %"class.sycl::_V1::ext::oneapi::device_global" zeroinitializer, align 8 #[[SAME_NAME_ANON_NS_ATTRS:[0-9]+]]
 
+struct bar {
+  int t1;
+  int t2;
+  [[intel::fpga_register]] int arr[155];
+};
+
+device_global<bar> counter;
+
+struct bar1 {
+  int t3;
+  int t4;
+  [[intel::max_replicates(2)]] int arr1[155];
+};
+
+device_global<bar1> counter1;
+
+struct bar2 {
+  int t5;
+  int t6;
+  [[intel::bankwidth(2)]] int arr2[155];
+};
+
+device_global<bar2> counter2;
+
+struct bar3 {
+  int t7;
+  int t8;
+  [[intel::fpga_memory]] int arr3[155];
+};
+
+device_global<bar3> counter3;
+
+struct bar4 {
+  int t9;
+  int t10;
+  [[intel::numbanks(2)]] int arr4[155];
+};
+
+device_global<bar4> counter4;
+
+struct bar5 {
+  int t11;
+  int t12;
+  [[intel::force_pow2_depth(1)]] int arr5[155];
+};
+
+device_global<bar5> counter5;
+
+struct bar6 {
+  int t13;
+  int t14;
+  [[intel::bank_bits(3, 4)]] int arr6[155];
+};
+
+device_global<bar6> counter6;
+
+struct bar7 {
+  int t15;
+  int t16;
+  [[intel::doublepump]] int arr7[155];
+};
+
+device_global<bar7> counter7;
+
+struct bar8 {
+  int t17;
+  int t18;
+  [[intel::singlepump]] int arr8[155];
+};
+
+device_global<bar8> counter8;
+
+struct bar9 {
+  int t19;
+  int t20;
+  [[intel::merge("mrg5", "width")]] int arr9[155];
+};
+
+device_global<bar9> counter9;
+
+struct bar10 {
+  int t21;
+  int t22;
+  [[intel::private_copies(8)]] int arr10[155];
+};
+
+device_global<bar10> counter10;
+
+struct bar11 {
+  int t23;
+  int t24;
+  [[intel::simple_dual_port]] int arr11[155];
+};
+
+device_global<bar11> counter11;
+
+struct bar12 {
+  [[intel::simple_dual_port]] int t25;
+  [[intel::private_copies(16)]] int t26;
+};
+
+device_global<bar12> counter12;
+
+struct bar13 {
+  [[intel::numbanks(2)]] int t27;
+  [[intel::merge("foo", "depth")]] int arr12[155];
+};
+
+device_global<bar13> counter13;
+
+struct bar14 {
+  [[intel::force_pow2_depth(0)]] int t28;
+  [[intel::doublepump]] int t29;
+  [[intel::bank_bits(2, 3)]] int arr13[155];
+};
+
+device_global<bar14> counter14;
+
+struct bar15 {
+  int t30;
+  int t31;
+  int arr14[155];
+};
+
+[[intel::fpga_register]] device_global<bar15> counter15;
```

Comment:
We should have test cases of derived classes with the bases containing fields that have different attributes on them, fields that are class types with attributes on them, and those classes containing fields that have attributes on them.

Reply:
Done